### DISCUSSION TITLE
error: no match for call to '(std::vector<double>) (libMesh::Real)'

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16139] [Q&A Modules: Phase field] [answered]
Hello guys,
I try to add a vector _M whose components are constant.
In the constructor function,
    _M(_etas_num*_etas_num, 0.0),
  for (unsigned int i = 0; i < _etas_num; ++i)
  {
    for (unsigned int j = 0; j < _etas_num; ++j)
    {
      if (0<i && i<_etas_num && j==_etas_num)
        _M(i*_etas_num+j)=1e-6;
      else if (0<j && j<_etas_num && i==_etas_num)
        _M(i*_etas_num+j)=1e-6;
      else
        _M(i*_etas_num+j)=7e-8;
    }
  }
In the .h file,
  std::vector<Real> _M;

It shows such error: 
**ACMultiR1.C:39:25: error: no match for call to '(std::vector<double>) (libMesh::Real)'
         _M(i*_etas_num+j)=1e-6;**
And another question:
**ACMultiR1.C:39:25: error: no match for call to '(std::vector<double>) (libMesh::Real)'
         _M(i*_etas_num+j)=1e-6;**

I have tried to correct it, but still failed. Could anyone help?
[ACMultiR1.zip](https://github.com/idaholab/moose/files/5517865/ACMultiR1.zip)

**Comment by aeslaughter** [comment]
> In the .h file,
> std::vector _M;

This is not the correct definition of a vector: https://en.cppreference.com/w/cpp/container/vector

I suggest you take some time to learn C++, we have some links on this page to some nice resources: https://mooseframework.inl.gov/getting_started/examples_and_tutorials/

**Reply by LiuPengPeter** [reply]
Thank you for your reply.
Sorry. In my question, I ignored it. But that's std::vector<Real> _M in the attachment.
After sleeping, I find the bug is that () was used when using vector's components...

There is still an error about another problem.
Could you help me check it? Bugs are so suffering...
**Second:** ACMultiR1.C:77:21: error: invalid types 'libMesh::Real {aka double}[unsigned int]' for array subscript  sum += 2/_Np[_qp]*_x[_index_var][_qp]*_x[j][_qp]*_M[_index_var*_etas_num+j][_qp]*_E[_index_var*_etas_num+j][_qp]/2

**Reply by aeslaughter** [reply]
```
 _M(i*_etas_num+j)=1e-6;
```
std::vectors use [] not (), which is very basic C++. Please spend some time going through some C++ tutorials.

**Reply by LiuPengPeter** [reply]
Yeah. That's very basic. I wrote () and [] in the same file...

I correct them, but there is still a problem about my Residual.

home/l/projects/park/src/Kernels/ACMultiR1.C:77:21: error: invalid types 'int[unsigned int]' for array subscript
     sum += 2/_Np[_qp]*_x[_index_var][_qp]*_x[j][_qp]*_M[_index_var*_etas_num+j][_qp]*_E[_index_var*_etas_num+j][_qp]/2*(_grad_u[_qp]-_grad_etas[j][_qp])*_grad_test[_i][_qp];
                     ^
/home/l/projects/park/src/Kernels/ACMultiR1.C:77:41: error: invalid types '__gnu_cxx::__alloc_traits<std::allocator<int> >::value_type {aka int}[unsigned int]' for array subscript
     sum += 2/_Np[_qp]*_x[_index_var][_qp]*_x[j][_qp]*_M[_index_var*_etas_num+j][_qp]*_E[_index_var*_etas_num+j][_qp]/2*(_grad_u[_qp]-_grad_etas[j][_qp])*_grad_test[_i][_qp];
                                         ^
/home/l/projects/park/src/Kernels/ACMultiR1.C:77:52: error: invalid types '__gnu_cxx::__alloc_traits<std::allocator<int> >::value_type {aka int}[unsigned int]' for array subscript
     sum += 2/_Np[_qp]*_x[_index_var][_qp]*_x[j][_qp]*_M[_index_var*_etas_num+j][_qp]*_E[_index_var*_etas_num+j][_qp]/2*(_grad_u[_qp]-_grad_etas[j][_qp])*_grad_test[_i][_qp];
                                                    ^
/home/l/projects/park/src/Kernels/ACMultiR1.C:77:84: error: invalid types '__gnu_cxx::__alloc_traits<std::allocator<double> >::value_type {aka double}[unsigned int]' for array subscript
     sum += 2/_Np[_qp]*_x[_index_var][_qp]*_x[j][_qp]*_M[_index_var*_etas_num+j][_qp]*_E[_index_var*_etas_num+j][_qp]/2*(_grad_u[_qp]-_grad_etas[j][_qp])*_grad_test[_i][_qp];
                                                                                    ^
/home/l/projects/park/src/Kernels/ACMultiR1.C:77:116: error: invalid types '__gnu_cxx::__alloc_traits<std::allocator<double> >::value_type {aka double}[unsigned int]' for array subscript
     sum += 2/_Np[_qp]*_x[_index_var][_qp]*_x[j][_qp]*_M[_index_var*_etas_num+j][_qp]*_E[_index_var*_etas_num+j][_qp]/2*(_grad_u[_qp]-_grad_etas[j][_qp])*_grad_test[_i][_qp];
                                                                                                                    ^
In file included from /home/l/projects/park/build/unity_src/Kernels_Unity.C:2:0:
/home/l/projects/park/src/Kernels/ACMultiR1.C:77:133: error: no match for 'operator-' (operand types are 'const libMesh::VectorValue<double>' and 'const MooseArray<libMesh::VectorValue<double> >')
[ACMultiR1.zip](https://github.com/idaholab/moose/files/5521291/ACMultiR1.zip)

**Comment by LiuPengPeter** [comment, marked as answer]
I delete some [_qp], then it works well.
